plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TemporalTimeTravelUtil.java

```diff
@@ -0,0 +1,292 @@
+            Optional<CommitInfoEntry> commitInfo = getEntriesFromJson(entryNumber, transactionLogDir, fileSystem, DataSize.ofBytes(0))
+                    .orElseThrow(() -> new MissingTransactionLogException(getTransactionLogJsonEntryPath(transactionLogDir, entryNumber).toString()))
+                    .getEntries(fileSystem)
+                    .map(DeltaLakeTransactionLogEntry::getCommitInfo)
+                    .filter(Objects::nonNull)
+                    .findFirst();
+            if (commitInfo.isPresent()) {
+                return searchBasedOnLatestCheckPoint(fileSystem, transactionLogDir, epochMillis, commitInfo.orElseThrow(), executor, maxLinearSearchSize);
+            }
+        }
+
+        if (fileSystem.newInputFile(getTransactionLogJsonEntryPath(transactionLogDir, 0)).exists()) {
+            return searchTowardsTail(fileSystem, 0, transactionLogDir, epochMillis, executor, maxLinearSearchSize);
+        }
+
+        return findLatestVersionFromWholeTransactions(fileSystem, transactionLogDir, epochMillis);
+    }
+
+    private static long searchBasedOnLatestCheckPoint(TrinoFileSystem fileSystem, String transactionLogDir, long epochMillis, CommitInfoEntry commitInfo, Executor executor, int maxLinearSearchSize)
+            throws IOException
+    {
+        long commitTime = getCommitInfoTimestamp(commitInfo);
+        if (commitTime == epochMillis) {
+            return commitInfo.version();
+        }
+
+        if (commitTime < epochMillis) {
+            long tail = searchTowardsTail(fileSystem, commitInfo.version() + 1, transactionLogDir, epochMillis, executor, maxLinearSearchSize);
+            if (tail >= 0) {
+                return tail;
+            }
+            return commitInfo.version();
+        }
+
+        return searchTowardsHead(fileSystem, commitInfo.version() - 1, transactionLogDir, epochMillis, executor, maxLinearSearchSize);
+    }
+
+    private static long searchTowardsHead(TrinoFileSystem fileSystem, long entryNumber, String transactionLogDir, long epochMillis, Executor executor, int maxLinearSearchSize)
+            throws IOException
+    {
+        if (entryNumber >= maxLinearSearchSize) {
+            return searchTowardsHeadParallel(fileSystem, entryNumber, transactionLogDir, epochMillis, executor, maxLinearSearchSize);
+        }
+
+        return searchTowardsHeadLinear(fileSystem, 0, entryNumber, transactionLogDir, epochMillis);
+    }
+
+    private static long searchTowardsTail(TrinoFileSystem fileSystem, long entryNumber, String transactionLogDir, long epochMillis, Executor executor, int maxLinearSearchSize)
+            throws IOException
+    {
+        if (fileSystem.newInputFile(getTransactionLogJsonEntryPath(transactionLogDir, entryNumber + maxLinearSearchSize)).exists()) {
+            return searchTowardsTailParallel(fileSystem, entryNumber, transactionLogDir, epochMillis, executor, maxLinearSearchSize);
```

Comment:
This likely does not make much sense at the moment.
If we'd have thousands of transaction log files after the checkpoint, this would rather be a design issue on the table definition.

Reply:
I was thought the case: the long exists table that the _last_checkpoint file happens to be deleted

Reply:
While this can theoretically happen,  this should not be happening in practice.

Reply:
Ok, let me remove it.